Branch of fesom2.6 including recom and tracer parallelisation

a270105 opened the pull request:
Tracer parallelisation can be switched on by __usetp in fesom2/src/CMakeLists.txt and onlye used when recom is on.

ogurses committed:
feat(fesom) add tracer diagnostics to FESOM2.5

Define new variables to track tracer changes
due to advection and diffusion.

We want to save for now diffusion and advection
contribution to the tracer changes. Horizontal and
vertical diffusion includes Redi
parametrization (if it is set .true.).

ogurses committed:
add recom_sms diag

ogurses committed:
feat(recom): add ciso and medusa components

Fill __ciso directive to ensure that
carbon isotope code works. Medusa interface is
added.

a270105 committed:
input to bc_surface from medusa added and Sed_input deleted

a270105 committed:
changes in src for tracer parallelisation

a270105 committed:
change defined(__recom)||defined(__usetp) to only defined(__usetp) in io_restart.F90

a270105 committed:
corrections to compile with tracer parallelisation on

a270105 committed:
cleanup and small corrections of flags for recom and tracer parallelisation

a270105 commented:
By compilation I found an issue regrading new compilations. I need to manually delete the executable file fesom.x in fesom2.build/bin and then compile the model. Otherwise, after a successful compilation I will still find the old fesom.x in fesom2/bin.

a270105 committed:
corrections in array allocation for Diags and initialization for use_atbox

JanStreffing reviewed (dismissed):
Summary of review:
There were a few places that need changing, some that may not need to change but would require a closer look that everything is indeed ok. There are a number of style improvements to be made. All comment that start with `! kh date` should start without the name and date. Same for `!YY`, `!O:G` `!OG` and `!.OG.` Git blame can tell you this info.

In addition to the standalone FESOM2 CI tests, this branch needs to be tested by me as part of AWI-CM3, by @ackerlar as part of AWI-ESM2, by @sebastianbeyer or @suvarchal as part of IFS-FESOM, by @mbutzin with active tracers. I'm assuming here, that you have already tested this branch with recom and without `_usetp`

JanStreffing commented:
In addition to the changes of the review, the merge conflicts should be solved. Most of them look like simple double additions of new features that both have to be kept. e.g. parallel tracers and icebergs.

ackerlar commented:
Is there any template fesom-recom yaml that I can use for testing?

a270105 commented:
> I get this error during compilation:
> 
> ```
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_mesh_ass.h(57): warning #5117: Bad # preprocessor line
> #if defined(__async_icebergs)
> -^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_mesh_ass.h(59): warning #5117: Bad # preprocessor line
> #endif
> -^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/cpl_driver.F90(363): error #6784: The number of actual arguments cannot be greater than the number of dummy arguments.   [OASIS_INIT_COMP]
>     CALL oasis_init_comp(comp_id, comp_name, ierror, num_program_groups = num_fesom_groups)
> ---------^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/cpl_driver.F90(961): error #6404: This name does not have a type, and must have an explicit type.   [MPI_COMM_FESOM_SAME_RANK_IN_GROUPS]
>       call MPI_Bcast(action, 1, MPI_LOGICAL, 0, MPI_COMM_FESOM_SAME_RANK_IN_GROUPS, MPIerr)
> ------------------------------------------------^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/cpl_driver.F90(961): error #6404: This name does not have a type, and must have an explicit type.   [MPIERR]
>       call MPI_Bcast(action, 1, MPI_LOGICAL, 0, MPI_COMM_FESOM_SAME_RANK_IN_GROUPS, MPIerr)
> ------------------------------------------------------------------------------------^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/cpl_driver.F90(976): error #6404: This name does not have a type, and must have an explicit type.   [MYDIM_NOD2D]
>           call MPI_Bcast(data_array, myDim_nod2d, MPI_DOUBLE_PRECISION, 0, MPI_COMM_FESOM_SAME_RANK_IN_GROUPS, MPIerr)
> -------------------------------------^
> compilation aborted for /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/cpl_driver.F90 (code 1)
> make[2]: *** [src/CMakeFiles/fesom.dir/build.make:205: src/CMakeFiles/fesom.dir/cpl_driver.F90.o] Error 1
> ```

I just compiled it with FESOM_COUPLED ON and didn't get any error. Did you compile the model with esm-tools?

JanStreffing commented:
Could be different compilers as well, some are more, some less strict on mismatching variable types.

ackerlar commented:
I got at least rid of this error
```
/home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/cpl_driver.F90(363): error #6784: The number of actual arguments cannot be greater than the number of dummy arguments.   [OASIS_INIT_COMP]
    CALL oasis_init_comp(comp_id, comp_name, ierror, num_program_groups = num_fesom_groups)
``` when switching to oasis branch `feat/multi-group-support`. Before I used `2.8mct-awiesm-2.1`. However, the other errors remain

ackerlar commented on src/fesom_module.F90:
```diff
@@ -147,8 +163,13 @@ subroutine fesom_init(fesom_total_nsteps)
-        call cpl_oasis3mct_init(f%partit,f%partit%MPI_COMM_FESOM)
+!        call cpl_oasis3mct_init(f%partit,f%partit%MPI_COMM_FESOM)
+! kh 02.12.21
+#if defined(__recom) && defined(__usetp)
```
"localCommunicator" is neither defined in fesom_module nor in MOD_PARTIT. Should this really be "localCommunicator" or "f%partit%MPI_COMM_FESOM"?

a270105 replied:
I can't really answer this. These lines already exited in fesom2.5-recom and another earlier version where Özgür and I merged their complex recom version and my simpler paleo version. But I can't find them in my paleo version where Kai first time implemented tracer parallelisation. There this part of code is still in fvom_main.f90 and only contains:
#if defined (__oasis)

! kh 21.03.22 pass num_fesom_groups to coupler
    call cpl_oasis3mct_init(MPI_COMM_FESOM, num_fesom_groups)
#endif
    t1 = MPI_Wtime()

    call par_init
......

So I think it should be for the case with tracer parallelisation:
call cpl_oasis3mct_init(f%partit, f%partit%MPI_COMM_FESOM, num_fesom_groups)

ackerlar committed:
make MPI calls consistent fix?? cpl_oasis3mct_init call in fesom_module

ackerlar commented:
FESOM compiled for me with some minor changes in MPI calls. I also changed

```
-        call cpl_oasis3mct_init(f%partit, f%partit%localCommunicator, num_fesom_groups)
+        call cpl_oasis3mct_init(f%partit, f%partit%MPI_COMM_FESOM, num_fesom_groups)
```
in `fesom_module` as `localCommunicator` is not defined outside of `cpl_driver`. Please check whether this makes sense.

However, the model crashes as FESOM is missing some entries in namelist.config. Is there a recom specific namelist.config?

a270105 commented:
> @a270105 yes, I used esm_tools: `esm_master get-awiesm-2.6`, changed the FESOM branch to `fesom2.6_recom_tp` and then `esm_master comp-awiesm-2.6`
> 
> which oasis version are you using?

a270105 commented:
> FESOM compiled for me with some minor changes in MPI calls. I also changed
> 
> ```
> -        call cpl_oasis3mct_init(f%partit, f%partit%localCommunicator, num_fesom_groups)
> +        call cpl_oasis3mct_init(f%partit, f%partit%MPI_COMM_FESOM, num_fesom_groups)
> ```
> 
> in `fesom_module` as `localCommunicator` is not defined outside of `cpl_driver`. Please check whether this makes sense.
> 
> However, the model crashes as FESOM is missing some entries in namelist.config. Is there a recom specific namelist.config?

Do you compile with __usetp? If not, there is no additional entry needed in namelist.config.

a270105 commented on src/io_restart.F90:
```diff
@@ -252,28 +270,56 @@ subroutine restart(istep, nstart, ntotal, l_read, which_readr, ice, dynamics, tr
   logical, save :: initialized_bin = .false.
-  integer mpierr
-  
+!  integer mpierr
```
No, since only partit%mpierr is used here

a270105 commented on src/oce_ale_tracer.F90:
```diff
@@ -270,10 +389,93 @@ subroutine solve_tracers_ale(ice, dynamics, tracers, partit, mesh)
         call exchange_nod(tracers%data(tr_num)%values(:,:), partit)
 !$OMP BARRIER
-    end do
+!    end do
```
It is not needed here but moved to L424.

JanStreffing replied:
We checked, and found that this is a shared end do between TP and non TP version later on in the code.

a270105 committed:
localcommunicator removed in fesom_module.F90

a270105 committed:
Merge branch 'fesom2.6_recom_tp' of https://github.com/FESOM/fesom2 into fesom2.6_recom_tp

ackerlar commented:
> Do you compile with __usetp? If not, there is no additional entry needed in namelist.config.

Yes, I set `RECOM_COUPLED=ON` which gives

```
if(${RECOM_COUPLED})
#   target_compile_definitions(${PROJECT_NAME} PRIVATE __recom USE_PRECISION=2 __usetp)
#   target_compile_definitions(${PROJECT_NAME} PRIVATE __recom USE_PRECISION=2)
   target_compile_definitions(${PROJECT_NAME} PRIVATE __recom USE_PRECISION=2 __3Zoo2Det __coccos __usetp)
endif()
``` 
in `src/CMakeLists.txt` and `__usetp` should be used if I understand correctly.

ackerlar commented:
Without `__usetp` I get these errors during compilation:

```
/home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_mesh_ass.h(59): warning #5117: Bad # preprocessor line
#endif
-^
/home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(2): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_WORLD]
MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
----------------------------------------------^
/home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(2): error #6795: The target must be of the same type and kind type parameters as the pointer.   [MPI_COMM_FESOM_WORLD]
MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
^
/home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(3): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_SAME_RANK_IN_GROUPS]
MPI_COMM_FESOM_SAME_RANK_IN_GROUPS  => partit%MPI_COMM_FESOM_SAME_RANK_IN_GROUPS
----------------------------------------------^
/home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(3): error #6795: The target must be of the same type and kind type parameters as the pointer.   [MPI_COMM_FESOM_SAME_RANK_IN_GROUPS]
MPI_COMM_FESOM_SAME_RANK_IN_GROUPS  => partit%MPI_COMM_FESOM_SAME_RANK_IN_GROUPS
^
/home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(2): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_WORLD]
MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
----------------------------------------------^
/home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(2): error #6795: The target must be of the same type and kind type parameters as the pointer.   [MPI_COMM_FESOM_WORLD]
MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
^
/home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(3): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_SAME_RANK_IN_GROUPS]
MPI_COMM_FESOM_SAME_RANK_IN_GROUPS  => partit%MPI_COMM_FESOM_SAME_RANK_IN_GROUPS
----------------------------------------------^
/home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(3): error #6795: The target must be of the same type and kind type parameters as the pointer.   [MPI_COMM_FESOM_SAME_RANK_IN_GROUPS]
MPI_COMM_FESOM_SAME_RANK_IN_GROUPS  => partit%MPI_COMM_FESOM_SAME_RANK_IN_GROUPS
```

patrickscholz commented:
> Without `__usetp` I get these errors during compilation:
> 
> ```
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_mesh_ass.h(59): warning #5117: Bad # preprocessor line
> #endif
> -^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(2): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_WORLD]
> MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
> ----------------------------------------------^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(2): error #6795: The target must be of the same type and kind type parameters as the pointer.   [MPI_COMM_FESOM_WORLD]
> MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
> ^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(3): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_SAME_RANK_IN_GROUPS]
> MPI_COMM_FESOM_SAME_RANK_IN_GROUPS  => partit%MPI_COMM_FESOM_SAME_RANK_IN_GROUPS
> ----------------------------------------------^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(3): error #6795: The target must be of the same type and kind type parameters as the pointer.   [MPI_COMM_FESOM_SAME_RANK_IN_GROUPS]
> MPI_COMM_FESOM_SAME_RANK_IN_GROUPS  => partit%MPI_COMM_FESOM_SAME_RANK_IN_GROUPS
> ^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(2): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_WORLD]
> MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
> ----------------------------------------------^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(2): error #6795: The target must be of the same type and kind type parameters as the pointer.   [MPI_COMM_FESOM_WORLD]
> MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
> ^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(3): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_SAME_RANK_IN_GROUPS]
> MPI_COMM_FESOM_SAME_RANK_IN_GROUPS  => partit%MPI_COMM_FESOM_SAME_RANK_IN_GROUPS
> ----------------------------------------------^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(3): error #6795: The target must be of the same type and kind type parameters as the pointer.   [MPI_COMM_FESOM_SAME_RANK_IN_GROUPS]
> MPI_COMM_FESOM_SAME_RANK_IN_GROUPS  => partit%MPI_COMM_FESOM_SAME_RANK_IN_GROUPS
> ```

I found some similar issues, when i was playing around on juwels last week with GCC compiler. There were different locations in the code where we tried to associate a pointer to a variable via the associate_part_ass.h and associate_mesh_ass.h file that wasnt allocated yet. The  happend especially during the initialisation phase! You can try to use the extended debuggin compiler flags and also the namelist.config flag_debug=.True., to better track down where it happens.
I solved my issue by checking in associate_part_ass.h if the variable is already allocated befor the pointer is setted
```
if (allocated(partit%remPtr_nod2D)) then
   remPtr_nod2D  (1:npes)                => partit%remPtr_nod2D(:)
end if
```

Commit:
set code version without tracer parallelisation as default

Commit:
set recom_coupled off as default

a270105 commented:
> Is there any template fesom-recom yaml that I can use for testing?

I created a new branch of esm-tools and added a new model 'awiesm-2.6-recom'. The branch is called `awiesm-2.6-recom` as well. 

In config/couplings I included `oasis3mct-2.8-paleodyn-multi-group-support` in the components instead of `oasis3mct-2.8-paleodyn` (for awiesm-2.6). Otherwise it does not differ from awiesm-2.6 except for awiesm.yaml, fesom-2.6.yaml and recom.yaml. I can install (and compile) the model and still need to work on a runscript. 
The relevant fesom namelists can be found in awiesm-2.6-recom/fesom-2.6/config/bin_2p1z1d. After testing I will put them into the default namelist dir.

a270105 commented:
> Without `__usetp` I get these errors during compilation:
> 
> ```
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_mesh_ass.h(59): warning #5117: Bad # preprocessor line
> #endif
> -^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(2): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_WORLD]
> MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
> ----------------------------------------------^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(2): error #6795: The target must be of the same type and kind type parameters as the pointer.   [MPI_COMM_FESOM_WORLD]
> MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
> ^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(3): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_SAME_RANK_IN_GROUPS]
> MPI_COMM_FESOM_SAME_RANK_IN_GROUPS  => partit%MPI_COMM_FESOM_SAME_RANK_IN_GROUPS
> ----------------------------------------------^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(3): error #6795: The target must be of the same type and kind type parameters as the pointer.   [MPI_COMM_FESOM_SAME_RANK_IN_GROUPS]
> MPI_COMM_FESOM_SAME_RANK_IN_GROUPS  => partit%MPI_COMM_FESOM_SAME_RANK_IN_GROUPS
> ^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(2): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_WORLD]
> MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
> ----------------------------------------------^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(2): error #6795: The target must be of the same type and kind type parameters as the pointer.   [MPI_COMM_FESOM_WORLD]
> MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
> ^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(3): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_SAME_RANK_IN_GROUPS]
> MPI_COMM_FESOM_SAME_RANK_IN_GROUPS  => partit%MPI_COMM_FESOM_SAME_RANK_IN_GROUPS
> ----------------------------------------------^
> /home/a/a270124/model_codes/testing_and_debugging/awiesm-2.6/fesom-2.6/src/associate_part_ass.h(3): error #6795: The target must be of the same type and kind type parameters as the pointer.   [MPI_COMM_FESOM_SAME_RANK_IN_GROUPS]
> MPI_COMM_FESOM_SAME_RANK_IN_GROUPS  => partit%MPI_COMM_FESOM_SAME_RANK_IN_GROUPS
> ```

Now I got these errors by compiling awiesm-2.6-recom. flag_debug is set to true and I didn't get more information.
These pointers are defined in associate_part_def.h when __recom and __usetp are both defined which happens through /src/CMakeLists. So with the default CMakeLists __usetp should not be defined and used at all. My suspicion is that the flag __usetp in /src/CMakeLists doesn't work. There is no problem with compilation and simulation on albedo with the ocean-only fesom-2.6-recom.

a270105 committed:
clean up kh comments

a270105 committed:
Merge branch 'fesom2.6_recom_tp' of https://github.com/FESOM/fesom2 into fesom2.6_recom_tp
update fesom-2.6

ackerlar committed:
point only when defined

ackerlar committed:
add missing tracer

ackerlar commented:
I was able to compile with these changes: [https://github.com/FESOM/fesom2/commit/4a053a9a234c1b1e5f2941007c0b4895962a8a74](https://github.com/FESOM/fesom2/commit/4a053a9a234c1b1e5f2941007c0b4895962a8a74)
and [https://github.com/FESOM/fesom2/commit/45d7fd33ac1aeb302021a6fde1eded481faef47e](https://github.com/FESOM/fesom2/commit/45d7fd33ac1aeb302021a6fde1eded481faef47e)

JanStreffing committed:
Merge branch 'main' into fesom2.6_recom_tp

JanStreffing committed:
remove spurious leftover end if

JanStreffing commented:
@mbutzin, please check out the two different ways to compute c14. Do we need both?

https://github.com/FESOM/fesom2/blob/4e0a621625a16a3d67b441010a99482381064df3/src/oce_ale_tracer.F90#L359-L371

JanStreffing committed:
try different preprocessor flags in .h file, hopefully passing CI

a270105 committed:
remove kh comments and fix for partit

a270105 commented:
I tried to compile the latest code in the branch `fesom2.6_recom_tp` on levante and still got this error:
`/work/ab0995/a270105/esm_code/awiesm-2.6-recom/fesom-2.6/src/associate_part_ass.h(2): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_WORLD]
MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD`
These preprocessor flags do not work on levante.

a270105 commented:
After installing cmake in my home directory on albedo the code of `fesom2.6_recom_tp` can be compiled with RECOM switched on/off and usetp switched on/off.

a270105 commented:
I am on the first task: fesom-recom without TP.
There is no error massage in slurm error report but fesom.out shows:
` ^[[36m     --> call vert_vel_ale^[[0m
  --> subroutine vert_vel_ale --> found Nan in Wvel after cumulativ summation(...)
  mype =          22
  node =           1
  Wvel(nzmin:nzmax, n)=                     NaN                     NaN                     NaN                     NaN                     NaN `
  Did we make something wrong in oce_ale.F90 by merging?

a270105 committed:
fix for ltra_diag in recom_main and for restart with TP in io_restart.F90.F90

a270105 committed:
set ltra_diag by default to false and improve alignment in fesom_module.F90

a270105 committed:
Merge branch 'main' into fesom2.6_recom_tp

a270105 committed:
fix in io_restart.F90 for writing restart

a270105 committed:
Merge branch 'fesom2.6_recom_tp' of github.com:FESOM/fesom2 into fesom2.6_recom_tp

a270105 committed:
fix continued

JanStreffing commented:
> > Great, next would be:
> > ```
> > * [x]  Checking that FESOM/RECOM run without TP
> > 
> > * [ ]  Checking that FESOM/RECOM run with TP
> > 
> > * [ ]  Checking that AWI-ESM2 & AWI-ESM3 compile without RECOM without TP
> > ```
> 
> I can't run with TP, since the code related to IO was changed a lot from fesom2.1 to fesom2.6. There are files that do not exist in fesom2.1. It would be helpful if somebody can explain to me where more changes are needed (for collecting tracers from parallel groups and writing into restart).

Can you have

a270105 committed:
fix in io_restart.F90

JanStreffing committed:
adapt recom restart writing for AB_order

a270105 committed:
index of data corrected

a270105 committed:
line with valuesold in io_restart commented

a270105 committed:
delete duplicates in io_restart

JanStreffing committed:
Merge branch 'main' into fesom2.6_recom_tp

JanStreffing committed:
Fix FESOM-2.6 compilation errors: MPI communicator structure members and duplicate SAVE attributes

- Remove conditional compilation for MPI_COMM_FESOM_WORLD, MPI_COMM_FESOM_SAME_RANK_IN_GROUPS, and my_fesom_group in MOD_PARTIT.F90
- Update associate_part_def.h to make MPI communicator pointers always available
- Update associate_part_ass.h to remove conditional compilation blocks
- Fix duplicate SAVE attribute warnings in cpl_driver.F90 by removing redundant save keywords

Fixes compilation errors where __recom and __usetp preprocessor flags were not defined during build.

a270105 committed:
include netcdf.inc added for reading sediment input file

JanStreffing commented:
Clear from my side. Maybe we should test it still as part of IFS-FESOM @suvarchal, @dsidoren, @sebastianbeyer.

suvarchal committed:
Merge branch 'main' into fesom2.6_recom_tp

suvarchal committed:
Fix preprocessor directives

fixed missing #endif after merge with profiling

JanStreffing committed:
Merge branch 'main' into fesom2.6_recom_tp

suvarchal reviewed (changes requested):
This is a very feature rich PR, i didn't manage to review till the end. While i don't understand well the rationales if tracer parallelization is prerequisite of recom and vice versa. I wonder this is in fact 2 rich features to fesom: recom and tracer parallelization. I think it would be more maintainable and testable code if it can be seperated to 2 PRs one for recom and other for tracer parallelization. The nested if defs for both features are hard to parse for my mere human abilities for time being and i am worried if it creates a lot of code-debt where alternative to separate them makes more sense to me.  Happy to have a call together with @JanStreffing and @patrickscholz to contribute to the effort if you are willing, i guess we need to take a day off and it is done.

a270105 commented:
@suvarchal I just talked to Jan. It makes sense to make tracer parallelisation a feature independent on recom if there are many other non-recom tracers to be calculated in parallel. Like I already explained, in Kai's implementation of tracer parallelisation in fesom2.1 we did not use any switch and the parallelisation was controlled by the number of tracer groups (one single parameter in namelist.config). If we should go for this option, you can first try the code without any __usetp.

Commit:
delete redundant comments in CMakelists

JanStreffing commented:
Given no changes, I think it's unrealistic to split recom from TP in this PR. I suggest we test what we have and if AWI-CM3, AWI-ESM2 and IFS-FESOM all work with this branch, then we add CI for RECOM: https://github.com/FESOM/fesom2/issues/620 and merge.

Commit:
add namelists to use tracer parallelisation

Commit:
Merge branch 'fesom2.6_recom_tp' of https://github.com/FESOM/fesom2 into fesom2.6_recom_tp

Commit:
change sizeof(cpl_send/recv) to nsend/nrecv in fesom_module.F90

a270105 commented:
The error does not occur with my changes in fesom_module.F90. But I still can't compile:

make[2]: *** [src/CMakeFiles/fesom.dir/build.make:2731: lib/libfesom.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:173: src/CMakeFiles/fesom.dir/all] Error 2
make: *** [Makefile:136: all] Error 2

What could cause this error?

a270105 commented:
It started with:
```
ld: /work/ab0995/a270105/esm_code/awiesm-2.6-recom/oasis/build/lib/psmile/libpsmile.a(mod_oasis_getput_interface.F90.o): relocation R_X86_64_32 against `__STRLITPACK_135' can not be used when making a shared object; recompile with -fPIC
ld: /work/ab0995/a270105/esm_code/awiesm-2.6-recom/oasis/build/lib/psmile/libpsmile.a(mod_oasis_grid.F90.o): relocation R_X86_64_32 against `__STRLITPACK_166' can not be used when making a shared object; recompile with -fPIC
ld: /work/ab0995/a270105/esm_code/awiesm-2.6-recom/oasis/build/lib/psmile/libpsmile.a(mod_oasis_io.F90.o): relocation R_X86_64_32 against `__STRLITPACK_1005' can not be used when making a shared object; recompile with -fPIC
```
...
and ends with:
```
ld: /work/ab0995/a270105/esm_code/awiesm-2.6-recom/oasis/build/lib/psmile/scrip/libscrip.a(remap_write.f.o): relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
ld: /work/ab0995/a270105/esm_code/awiesm-2.6-recom/oasis/build/lib/psmile/scrip/libscrip.a(netcdf.F.o): relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
```

JanStreffing commented:
Can you change the comp_command in esm_tools to include `-fPIC?` as suggested? Maybe that already helps.

JanStreffing commented:
> Can you change the comp_command in esm_tools to include `-fPIC?` as suggested? Maybe that already helps.

Any luck with this one @a270105 ?

Commit:
remove usetp from default to test awiesm-2.6-recom

a270105 commented:
> 
>     * [ ]  AWI-ESM2, without RECOM, without TP [@a270105  ]
> 
can't do this. Either with or without recom can't not be compiled if coupled. FESOM only has no problem.

a270105 commented:
The problem is solved. But I need somebody to check if it is only a problem with my esm-tools branch.
I first tried with environment variable for oasis like `export FCFLAGS="-fPIC"`. It doesn't help. And in oasis3mct.env.yaml these are included:
`computer:
    choose_general.execution_mode:
        compile:
            add_export_vars:
                OASIS_FFLAGS: '"-fPIC"'
                OASIS_CFLAGS: '"-fPIC"'
                CFLAGS: '"-fPIC"'
                CCFLAGS: '"-fPIC"'
                FFLAGS: '"-fPIC"'
                FCFLAGS: '"-fPIC"'
`
Then I added these in the compilation shell script (comp-oasis3mct-2.8-paleodyn-multi-group-support_script.sh):
`-DCMAKE_POSITION_INDEPENDENT_CODE=ON -DCMAKE_Fortran_FLAGS="-fPIC"`
and recompiled oasis. After that I can compile fesom in the coupled system.
Can somebody explain this and how it is done for awiesm-2.6?

JanStreffing commented:
Adding 

> The problem is solved. But I need somebody to check if it is only a problem with my esm-tools branch. I first tried with environment variable for oasis like `export FCFLAGS="-fPIC"`. It doesn't help. And in oasis3mct.env.yaml these are included: `computer: choose_general.execution_mode: compile: add_export_vars: OASIS_FFLAGS: '"-fPIC"' OASIS_CFLAGS: '"-fPIC"' CFLAGS: '"-fPIC"' CCFLAGS: '"-fPIC"' FFLAGS: '"-fPIC"' FCFLAGS: '"-fPIC"' ` Then I added these in the compilation shell script (comp-oasis3mct-2.8-paleodyn-multi-group-support_script.sh): `-DCMAKE_POSITION_INDEPENDENT_CODE=ON -DCMAKE_Fortran_FLAGS="-fPIC"` and recompiled oasis. After that I can compile fesom in the coupled system. Can somebody explain this and how it is done for awiesm-2.6?

Adding some AWI-ESM2 users: @ackerlar, @PengyangSong, @christian-stepanek, @chrisdane

Commit:
add -fPIC in CMakeLists

Commit:
put num_fesom_groups back to run_config

Commit:
delete read_namelist_run_config

Commit:
turn on RECOM_COUPLED by default

JanStreffing commented:
Can you link to that branch? 
I can't see that online

a270105 commented:
https://gitlab.awi.de/paleodyn/Models/oasis3-mct/-/tree/feat/multi-group-support?ref_type=heads

a270105 commented:
yes, I also found his changes in these 3 files

Commit:
add read_namelist_run_config back to fesom_module.F90

JanStreffing commented:
> I can't run awiesm-2.6-recom with TP with the current oasis (from the branch 'multi-group-support') and got segmentation fault at the step `FESOM: calling exchange_roots`. So, at the moment the code works with TP compiled and num_fesom_groups = 1 (only one tracer groups, i.e. no MPI applied for tracers). For further steps I need help from Kai or someone else.

Good to know. This does not block the merge, as that would be a new feature.

JanStreffing commented:
> I tried the AWI-ESM3 part. I get stuck compiling:
> 
> ```
> /work/ab0246/a270092/model_codes/awiesm3-develop/fesom-2.6/src/cpl_driver.F90(362): error #6627: This is an actual argument keyword name, and not a dummy argument name.   [NUM_PROGRAM_GROUPS]
>     CALL oasis_init_comp(comp_id, comp_name, ierror, num_program_groups = num_fesom_groups)
> -----------------------------------------------------^
> compilation aborted for /work/ab0246/a270092/model_codes/awiesm3-develop/fesom-2.6/src/cpl_driver.F90 (code 1)
> make[2]: *** [src/CMakeFiles/fesom.dir/build.make:205: src/CMakeFiles/fesom.dir/cpl_driver.F90.o] Error 1
> make[2]: *** Waiting for unfinished jobs....
> ```
> 
> Does that look familiar to you @a270105 ?

I no longer get this error using the Group enabled version of OASIS3MCT5.
Next I got errors with preprocessing flags in the the .h files. Basically Intel compiler seems to ignore preprocessing flags inside .h files. When use_tp is not set, it then tries to access pieces of partit, that are not there, because the working preprocessor statements in the F90 code removed them:

```
/work/ab0246/a270092/model_codes/awiesm3-develop-cc/fesom-2.6/src/associate_part_ass.h(2): error #6460: This is not a component name that is defined in the encompassing structure.   [MPI_COMM_FESOM_WORLD]
MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
----------------------------------------------^
```

JanStreffing committed:
replace ifdef with DEC preprocessing flags in .h files

JanStreffing committed:
Merge branch 'fesom2.6_recom_tp' of ssh://github.com/FESOM/fesom2 into fesom2.6_recom_tp

a270105 commented:
I just looked back in this merge request. Yes, I had the same issue and Lars as well. I tried with !DIR since I thought it would be good to not to be compiler specific...But for some reasons it did not work. So we decided to remove the preprocessor directives and let those pointers always defined. The same is done for the iceberg part. I just have these lines in the beginnging of my associate_part_ass.h:
`MPI_COMM_FESOM_WORLD                => partit%MPI_COMM_FESOM_WORLD
MPI_COMM_FESOM_SAME_RANK_IN_GROUPS  => partit%MPI_COMM_FESOM_SAME_RANK_IN_GROUPS
MPI_COMM_FESOM  => partit%MPI_COMM_FESOM
MPI_COMM_FESOM_IB       => partit%MPI_COMM_FESOM_IB
`
Do you have a different file?

Commit:
remove preprocessor directives in .h files

JanStreffing commented:
Adding the tests to the end, so they are easier to find:


Here are the tests we need to pass once before we can merge this branch:

Compiling:
- [x] FESOM2, standalone, without RECOM, without TP [@a270105 ]
- [x] FESOM2, standalone, with RECOM, without TP [@a270105 ]
- [x] FESOM2, standalone, with RECOM, with TP [@a270105 ]
- [x] AWI-ESM2, without RECOM, without TP [@a270105  ]
- [x] AWI-ESM2, with RECOM, without TP [@a270105 ]
- [x] AWI-ESM2, with RECOM, with TP [@a270105 ] (nice to have but not blocking this PR if it does not work) 
- [x] AWI-CM3, without RECOM, without TP [@JanStreffing]
- [x] AWI-ESM3, with RECOM, without TP [@a270105]
- [x] AWI-ESM3, with RECOM, with TP [@a270105] (nice to have but not blocking this PR if it does not work)
- [ ] IFS-FESOM, without RECOM, without TP [@suvarchal]

Running for 1 day:
- [ ] FESOM2, standalone, without RECOM, without TP [@a270105 ]
- [x] FESOM2, standalone, with RECOM, without TP [@a270105 ]
- [x] FESOM2, standalone, with RECOM, with TP [@a270105 ]
- [ ] AWI-ESM2, without RECOM, without TP [@a270105 ]
- [x] AWI-ESM2, with RECOM, without TP [@a270105 ]
- [ ] AWI-ESM2, with RECOM, with TP [@a270105 ] (nice to have but not blocking this PR if it does not work) 
- [x] AWI-CM3, without RECOM, without TP [@JanStreffing]
- [ ] AWI-ESM3, with RECOM, without TP [@chrisdane]
- [ ] AWI-ESM3, with RECOM, with TP [@chrisdane] (nice to have but not blocking this PR if it does not work)
- [ ] IFS-FESOM, without RECOM, without TP [@suvarchal]

Update: `AWI-ESM3, with RECOM, without TP`  and `AWI-ESM3, with RECOM, with TP` compile.

a270105 committed:
pH/DOC dependent parameterisation of Fe-binding ligands completed

a270105 commented:
I have to report something really odd:
I added a new feature in the recom code (src/int_recom/recom_sms.F90) and recompiled fesom. However, a fully identical executable file was created (the build folder was deleted before compilation). This did not happen as I changed the fesom code (or cmake options in src/CMakeLists for recom). Somehow the recom files were not updated during recompilation. Others in the recom group have the same problem with other branches like fesom-2.6-recom-tra-diags. 
This might need to be fixed before the release, right?

Commit:
fix for tracer parallelisation in coupled setup

Commit:
Merge branch 'fesom2.6_recom_tp' of https://github.com/FESOM/fesom2 into fesom2.6_recom_tp

Commit:
correction in fesom_module.F90 to allocate nsend and nrecv for parallel groups

a270105 committed:
set flag_debug to false

a270105 committed:
merge conflict fix

a270105 committed:
bugfix in io_resatart.F90

a270105 committed:
bugfix in io_restart.F90 after merge

Commit:
bugfix in bio_files in io_restart.F90

a270105 committed:
fesom namelists updated in bin_2p1z1d_tp

Commit:
namelists updated for 2.7

Commit:
fix in depth denpendecy of calcite dissolution

Commit:
DIC_PI false by default

Commit:
Merge branch 'main' into fesom2.6_recom_tp

a270105 committed:
updating namelists and grouping iron chesmitry paramters in namelist.recom

JanStreffing committed:
Merge branch 'main' into fesom2.6_recom_tp

Commit:
Merge branch 'fesom2.6_recom_tp' of https://github.com/FESOM/fesom2 into fesom2.6_recom_tp
updating local branch

Commit:
bugfix for sedimentary input from MEDUSA in gen_surface_forcing.F90

Commit:
fix of line breaks in recom namelist

Commit:
delete namelists not needed for 2.7

Commit:
updating namelists

Commit:
changes by Kai for benthic and sinking fluxes when using tracer parallelisation, sinking fluxes needed for MEDUSA added in io_meandata.F90

Commit:
modification for biological fractionation of carbon

JanStreffing committed:
Merge branch 'main' into fesom2.6_recom_tp

JanStreffing committed:
Merge branch 'main' into fesom2.6_recom_tp

JanStreffing committed:
add partit arg to read_namelist_run_config call

Commit:
changes wrt. C isotopes

Commit:
fix to read num_fesom_groups

a270105 committed:
Merge branch 'main' into fesom2.6_recom_tp

suvarchal committed:
fix preprocessor oasis and tracer parallel block

suvarchal committed:
get rid of warnings of using comma before io list

JanStreffing committed:
Merge branch 'main' into fesom2.6_recom_tp

suvarchal commented:
> These tests all need to work, yes. @suvarchal, can you have a look with @a270105 ?

give me a day, i will fix these tests, we can hopefully merge then.

Commit:
add switch for variable grazing efficiency

Commit:
merging between all tracer groups immediately before writing restart

Commit:
diagnostic output pH and CO3 added

a270105 commented:
What is the status of this merge request? 
Does the main branch contain all updates from Ozgur (fesom2.7-recom3.1)? I see there are much more conflicts now...

a270105 committed:
sbc_do_recom added for a separte surface forcing for REcoM

a270105 committed:
add recom directive for sbc_do_recom

Commit:
sbc_ini_recom added to read nam_rsbc from namelist.recom

Commit:
reading of tp paramter improved

JanStreffing commented:
Superseeded by #976
